core/src/main/java/software/amazon/awssdk/pagination/PaginatedItemsIterable.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.pagination;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.function.Function;
+
+public class PaginatedItemsIterable<ResponseT, ItemT> implements SdkIterable<ItemT> {
+
+    private final Iterator<ResponseT> responseIterator;
+    private final Function<ResponseT, Iterator<ItemT>> getPaginatedItemIterator;
+
+    public PaginatedItemsIterable(Iterator<ResponseT> responseIterator,
+                                  Function<ResponseT, Iterator<ItemT>> getPaginatedItemIterator) {
+        this.responseIterator = responseIterator;
+        this.getPaginatedItemIterator = getPaginatedItemIterator;
+    }
+
+
+    @Override
+    public Iterator<ItemT> iterator() {
+        return new ItemIterator(responseIterator);
+    }
+
+    private class ItemIterator implements Iterator<ItemT> {
+
+        private Iterator<ItemT> itemsIterator;
+
+        ItemIterator(Iterator<ResponseT> responseIterator) {
+            this.itemsIterator = getPaginatedItemIterator.apply(responseIterator.next());
+        }
+
+        @Override
+        public boolean hasNext() {
+            return (itemsIterator != null && itemsIterator.hasNext()) ||
```

Comment:
For items iteration it's possible for the responseIterator to have a next page but that next page have an empty collection of items. We should make sure we are handling that case correctly in accordance with the Iterator interface.

Reply:
Is it possible for a page to have empty items? Then what is the use of that page? At this point in code, we will throw an exception as there is no next element. Its tricky to handle this case as we don't know if the page with empty items is the last page or not.

Reply:
Yeah it's possible, at least for Dynamo and I suspect for other services too. I think it's just an implementation limitation, the service may not know what the last item is until it goes to fetch it so if you end a page on the last item it thinks there might be more when in reality there's not. The only way to check for sure is to make another call and get an emtpy response with a empty paging token.

http://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html

> The primary key of the item where the operation stopped, inclusive of the previous result set. Use this value to start a new operation, excluding this value in the new request.
If LastEvaluatedKey is empty, then the "last page" of results has been processed and there is no more data to be retrieved.
If LastEvaluatedKey is not empty, it does not necessarily mean that there is more data in the result set. The only way to know when you have reached the end of the result set is when LastEvaluatedKey is empty.